error handling

Sam09040 opened the pull request:
## add
- custom error to handle the errors
- error test file with four tests
- script to choose which test file to run

## update
- error handling in resolvers
- test env file name

Sam09040 committed:
Inicial setup

- test file with user mutation in testdb functioning

- add app folder and put interfaces, client and graphql content there

Sam09040 committed:
small changes

- tsimp to run .ts
- run configuration with tsimp

- test file from .mjs to .ts
- resolvers import on index.ts

- .prettierrc.ts file as it was a duplicate

Sam09040 committed:
error handling

- custom error to handle the errors
- error test file with four tests
- script to choose which test file to run

- error handling in resolvers
- test env file name

racsandoval commented on run-script.ts:
```diff
@@ -0,0 +1,12 @@
+const args = process.argv.slice(2);
```
qual uso desse arquivo?

Sam09040 replied:
ele permite que eu escolha qual arquivo de teste eu quero rodar passando um argumento no cli. por exemplo, npm run test -- test/index.test.ts. eu achei interessante pois me permite rodar diferentes testes de forma mais dinâmica, sem ter que ir mudando no script de teste.

walteraandrade replied:
o mocha permite que você faça isso de forma mais prática, na minha opinião: https://mochajs.org/#exclusive-tests

walteraandrade commented on src/app/graphql/resolvers.ts:
```diff
@@ -16,19 +17,43 @@ export const resolvers = {
       return users;
     },
     user: async (_: unknown, { id }: { id: number }) => {
+        if(id == 0){
```
por que `==` ?

walteraandrade commented on src/app/graphql/resolvers.ts:
```diff
@@ -16,19 +17,43 @@ export const resolvers = {
-
-      if (!user) {
-        throw new Error('User not found!');
+      if (user == null) {
```
no geral, evitamos usar `==` em javascript, é uma ferramenta que o próprio time que mantém a linguagem admite ser muito aberta a bugs e comportamentos inesperados

SamuelSsc replied:
aqui ainda falta ver esse comentário do Waltinho, sobre o uso do `==`

walteraandrade commented on test/error.test.ts:
```diff
@@ -0,0 +1,175 @@
+            }
+        };
+        
+        console.log('Sending query...');
```
log

racsandoval commented on test/error.test.ts:
```diff
@@ -0,0 +1,175 @@
+        if(!prisma.$connect()){
+            await prisma.$connect();
+            console.log('Connected to testdb');
+        }
```
acho que aqui vale alterar para a forma como ficou no PR anterior

racsandoval commented on test/index.test.ts:
acho que `index` não é um bom nome pra esse arquivo

racsandoval commented on test/index.test.ts:
```diff
@@ -0,0 +1,90 @@
+
+        expect(userInDb).to.not.equal(null);
+        expect(userInDb?.name).to.equal('Sam');
+        console.log(userInDb?.name);
```
precisa desse log?

Sam09040 replied:
já apaguei!

sophia-bergamo commented on src/app/graphql/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+            'Invalid request!',
+            {
+              field: 'id',
+              reason: 'The id needs to be greater than 0!',
```
não sei se precisa verificar se o id é 0, acho que é mais seguro somente verificar se existe o id no banco e se não achar lança o erro de não encontrado. Assim, quem estiver tentando fazer a requisição com o id 0 não vai saber se esse id é válido ou não.

Sam09040 replied:
hmmm sim sim, é que eu estava tentando tratar todos os erros que eu conseguia tratar e esse de id sendo 0 era um deles, mas faz sentido o que você disse!

sophia-bergamo commented on src/app/graphql/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+      }
+
+      const existingEmail = await prisma.user.findUnique({
+        where: { email: email },
```
depois tira um tempinho pra estudar sobre `Object Shorthand Notation`

```suggestion
        where: { email },
```

Sam09040 replied:
sim sim!! obrigada pela sugestão!

sophia-bergamo commented on src/app/graphql/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+      if (!isPasswordValid(password)) {
+        throw new CustomError(
+          '401',
+          'Password doesn\'t fit requirements!',
```
```suggestion
          `Password doesn't fit requirements!`,
```

walteraandrade replied:
acho que fica mais legível também

Sam09040 replied:
inclusive, interessantemente, o prettier percebeu que eu tinha feito dessa forma e alterou de aspas simples para aspas duplas, eliminando a necessidade da \.

sophia-bergamo commented on test/index.test.ts:
```diff
@@ -0,0 +1,90 @@
+            variables,
+        });
+        const { data } = response.data;
+        console.log('response received: ', data);
```
log

Sam09040 committed:
changes post rebase + fixes

## updates:
- mocha runs all test files, but only executes the suites with .only();
- followed suggestions and made fixes accordingly;
- changed test file name from index to first;
- imports of server and prisma variables;

## deletes:
- run-script.ts, now using .only() to run specific tests;
- all logs that weren't necessary;

SamuelSsc commented on test/error.test.ts:
```diff
@@ -0,0 +1,177 @@
+      })
+    }
+  });
+})
```
nit: adiciona uma linha vazia aqui ao fim do arquivo

Sam09040 committed:
corrections

## updates:
- ran prettier through test files;
- changed == to ===;

racsandoval commented on test/error.test.ts:
```diff
@@ -0,0 +1,177 @@
+                    birthDate
+                }
+            }
+        `;
```
aqui essa variável continua se repetindo em todos os testes

Sam09040 committed:
small change on error.test.ts

## update:
- mutation variable declared only once;
- erased unecessary log;
